securedrop/encryption.py

```diff
@@ -0,0 +1,256 @@
+        """
+        Parse a status code from the attached GnuPG process.
+        :raises: :exc:`~exceptions.ValueError` if the status message is unknown.
+        """
+        if key in ("DELETE_PROBLEM", "KEY_CONSIDERED"):
+            self.status = self.problem_reason.get(value, "Unknown error: %r" % value)
+        elif key in ("PINENTRY_LAUNCHED"):
+            self.status = key.replace("_", " ").lower()
+        else:
+            raise ValueError("Unknown status message: %r" % key)
+
+    gnupg._parsers.DeleteResult._handle_status = _updated_handle_status
+
+
+def _setup_monkey_patches_for_gnupg() -> None:
+    _monkey_patch_username_in_env()
+    _monkey_patch_unknown_status_message()
+    _monkey_patch_delete_handle_status()
+
+
+_setup_monkey_patches_for_gnupg()
+
+
+class GpgKeyNotFoundError(Exception):
+    pass
+
+
+class GpgEncryptError(Exception):
+    pass
+
+
+class GpgDecryptError(Exception):
+    pass
+
+
+_default_encryption_mgr: Optional["EncryptionManager"] = None
+
+
+class EncryptionManager:
+
+    GPG_KEY_TYPE = "RSA"
+    GPG_KEY_LENGTH = 4096
+
+    # All reply keypairs will be "created" on the same day SecureDrop (then
+    # Strongbox) was publicly released for the first time.
+    # https://www.newyorker.com/news/news-desk/strongbox-and-aaron-swartz
+    DEFAULT_KEY_CREATION_DATE = date(2013, 5, 14)
+
+    # '0' is the magic value that tells GPG's batch key generation not
+    # to set an expiration date.
+    DEFAULT_KEY_EXPIRATION_DATE = "0"
+
+    REDIS_FINGERPRINT_HASH = "sd/crypto-util/fingerprints"
+    REDIS_KEY_HASH = "sd/crypto-util/keys"
+
+    SOURCE_KEY_NAME = "Source Key"
+    SOURCE_KEY_UID_RE = re.compile(r"(Source|Autogenerated) Key <[-A-Za-z0-9+/=_]+>")
+
+    def __init__(self, gpg_key_dir: Path, journalist_key_fingerprint: str) -> None:
+        self._gpg_key_dir = gpg_key_dir
+        self._journalist_key_fingerprint = journalist_key_fingerprint
+        self._redis = Redis(decode_responses=True)
+
+        # Instantiate the GPG binary
+        gpg = gnupg.GPG(binary="gpg2", homedir=str(self._gpg_key_dir))
+        if StrictVersion(gpg.binary_version) >= StrictVersion("2.1"):
+            # --pinentry-mode, required for SecureDrop on GPG 2.1.x+, was added in GPG 2.1.
+            self._gpg = gnupg.GPG(
+                binary="gpg2", homedir=str(gpg_key_dir), options=["--pinentry-mode loopback"]
+            )
+        else:
+            self._gpg = gpg
+
+        # Ensure that the journalist public key has been previously imported in GPG
+        try:
+            self.get_journalist_public_key()
+        except GpgKeyNotFoundError:
+            available_keys = self._gpg.list_keys()
+            raise EnvironmentError(
+                f"The journalist public key with fingerprint {journalist_key_fingerprint}"
+                f" has not been imported into GPG. "
+                f"Keys currently configured in GPG: {available_keys}"
```

Comment:
This last line will list reply keys in the logs, giving the number of active sources on the instance - said logs might be shared for troubleshooting purposes, so it might be better not to include it.